### PR TITLE
[ATOM][Kimi-K3] ATOM support Kimi-K3

### PR DESCRIPTION
**Pull request by zejunchen-zejun**

### CODE REVIEW

File: atom/model_engine/engine_core_mgr.py [hunk 1]
```diff
@@ -793,6 +792,7 @@ def launch_engine_core(config: Config, dp_rank: int = 0):
             "config": config,
             "input_address": input_address,
             "output_address": output_address,
+            "env_snapshot": dict(os.environ),
```
Comment (wuhuikx) [inline atom/model_engine/engine_core_mgr.py]: do we need this change?

File: atom/model_engine/model_runner.py [hunk 2]
```diff
@@ -753,8 +754,33 @@ def __init__(self, rank: int, config: Config):
...
+            config.load_dummy,
+            load_fused_expert_weights_fn=fused_shared_expert_load_fn,
+        )
+        load_elapsed = time.perf_counter() - load_start
+        logger.info(
+            f"[{self.rank_name}] Model load done: {config.model} "
+            f"(weights loaded in {load_elapsed:.2f}s)"
+        )
+        if (
+            os.getenv("ATOM_SYNC_AFTER_LOAD", "0").lower() in ("1", "true", "yes")
```
Comment (wuhuikx) [inline atom/model_engine/model_runner.py:776]: Please remove this

File: atom/model_engine/model_runner.py [hunk 3]
```diff
@@ -1213,6 +1354,16 @@ def warmup_model(self):
...
 
+        # TEMPORARY (benign, warmup-only): cap the dummy warmup prefill. A large
+        # all-zero warmup batch samples garbage logits over all positions and
+        # faults the sampler/softmax on gfx1250; real inference only samples the
+        # last token per seq so it is unaffected. Proper fix = skip/greedy sampling
+        # during is_dummy_run warmup, then this cap can go. Not needed for the MoE
+        # (that large-M crash is fixed in aiter grouped_moe_gfx1250).
+        _warmup_cap = int(os.environ.get("ATOM_WARMUP_MAX_TOKENS", "0") or "0")
+        if _warmup_cap > 0:
+            warmup_max_tokens = min(warmup_max_tokens, _warmup_cap)
```
Comment (wuhuikx) [inline atom/model_engine/model_runner.py:1365]: do we need this?

File: atom/model_loader/loader.py [hunk 4]
```diff
@@ -70,6 +147,175 @@ def default_weight_loader(param: nn.Parameter, loaded_weight: torch.Tensor):
...
+            pbar.update(1)
+            if enable_tqdm:
+                logger.info(
+                    "Finished safetensors shard %d/%d in %.2fs: %s",
+                    batch_start + 1,
+                    len(hf_weights_files),
+                    time.perf_counter() - shard_start,
+                    os.path.basename(st_file),
+                )
+    pbar.close()
```
Comment (wuhuikx) [inline atom/model_loader/loader.py:316]: Please check the above code.

File: atom/model_engine/engine_core_mgr.py (hunk 1 above)
Comment (zejunchen-zejun) [reply atom/model_engine/engine_core_mgr.py]: removed

File: atom/model_engine/scheduler.py [hunk 5]
```diff
@@ -1522,6 +1455,15 @@ def postprocess(
             num_placeholder += 1
 
         for seq in self.running:
+            # A disconnected client may abort a seq while its current forward is
+            # still in flight. Do not try to backfill deferred placeholders for
+            # that seq: aborted partial-prefill requests may not have any
+            # output_tokens placeholder to overwrite.
+            if seq.status == SequenceStatus.ABORTED:
```
Comment (valarLip) [inline atom/model_engine/scheduler.py:1462]: ABORTED should already supported? @yhl-amd any comments?

File: atom/entrypoints/openai/reasoning.py [hunk 6]
```diff
@@ -12,6 +12,22 @@
 from dataclasses import dataclass
 from typing import Optional, Tuple
 
+KIMI_THINK_END = "<|close|>think<|sep|>"
```
Comment (valarLip) [inline atom/entrypoints/openai/reasoning.py:15]: no model specific stuff should be here

File: atom/model_ops/moe.py [hunk 7]
```diff
@@ -71,15 +62,19 @@
...
 )
 from atom.utils import envs
 from atom.utils.custom_register import direct_register_custom_op
 from atom.utils.decorators import mark_trace
 from atom.utils.forward_context import get_forward_context
+from torch import nn
+from transformers import PretrainedConfig
 
 logger = logging.getLogger("atom")
+_MOE_LOAD_COPY_LOG_SECONDS = float(os.getenv("ATOM_MOE_LOAD_COPY_LOG_SECONDS", "0"))
```
Comment (valarLip) [inline atom/model_ops/moe.py:77]: ...

File: atom/model_ops/moe.py [hunk 8]
```diff
@@ -3004,8 +3054,31 @@ def _load_per_channel_weight_scale(
     @staticmethod
     def _copy_quant_storage(dst: torch.Tensor, src: torch.Tensor) -> None:
         """Copy quantized tensors without numeric conversion of byte formats."""
+        if src.device.type == "cpu" and not src.is_contiguous():
+            src = src.contiguous()
+        start = time.perf_counter() if _MOE_LOAD_COPY_LOG_SECONDS > 0 else None
+
+        def log_slow_copy() -> None:
```
Comment (valarLip) [inline atom/model_ops/moe.py:3061]: ...

File: atom/model_ops/sampler.py [hunk 9]
```diff
@@ -73,6 +74,11 @@ def forward(
         Returns:
             Sampled token IDs (num_tokens,)
         """
+        if envs.ATOM_DEBUG_TOPK > 0:
```
Comment (valarLip) [inline atom/model_ops/sampler.py:77]: ..

File: atom/models/kimi_k3_fused.py
Comment (valarLip) [inline atom/models/kimi_k3_fused.py:1]: most of kernels are duplicated with existing one..

File: atom/config.py [hunk 10]
```diff
@@ -634,9 +646,15 @@ def _get_hf_token() -> str | None:
         ):
             text_config_dict["quantization_config"] = config_dict["quantization_config"]
         text_model_type = text_config_dict.get("model_type", "deepseek_v3")
```
Comment (valarLip) [inline atom/config.py:648]: text_model_type = text_config_dict.get("model_type", "deepseek_v3") ->text_model_type = text_config_dict.get("model_type", "text_config")

**Comment by zejunchen-zejun** [comment]
use PR https://github.com/ROCm/ATOM/pull/1718
